### PR TITLE
Rewrite README

### PR DESCRIPTION
**Pull request by morrisallison**
A well needed update. Don't bother looking at the diff, just [read the actual document](https://github.com/yola/jsgreat/blob/morris/update/README.md).

To clarify; there should be no surprises in here, just updates to match our current style.

:bell: ping @yola/frontend-engineering

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,222 +1,515 @@
...
-## Using Variables Before Definition
+**Don't do this:**
 
-Variables must be defined earlier in the code than they are used. (Hoisting makes it possible to accidentally use a variable before it's defined.) This will be [enforced by JSHint](http://www.jshint.com/docs/options/#latedef).
+```javascript
+let longTestPermalinkUrl = 'https://example.com/v1/en/2016/03/14/assets/images/1234.jpg';
+```
 
-## Trailing Whitespace
+**This is allowed:**
```
Comment (Hkly) [inline README.md]: Do you wanna say "Do this instead" to conform with the rest of the doc?

Comment (morrisallison) [inline README.md]: Yeah, it doesn't make much sense in this spot. The "In specific cases" is enough.

**Comment by Hkly** [comment]
Could we add (I guess only if everyone agrees on this) to enforce parens with ES6 arrow functions, even if there is only one parameter?

File: README.md [hunk 2]
```diff
@@ -1,222 +1,515 @@
...
+
+```javascript
+let foo = function() {
+  /// [...]
+};
+```
+
+#### Anonymous Functions
+
+Function expressions should be used instead of anonymous functions where applicable.
```
Comment (DevShep) [inline README.md]: In regards to https://github.com/yola/jsgreat/issues/22, anonymous functions should be allowed. Some examples of when I think they're great: functional programming, `=>` functions that can be one line, anytime naming a function adds more lines of code without much if any meaning or context.

I do think that in promise chains, though, anonymous functions should not be allowed.

File: README.md [hunk 3]
```diff
@@ -1,222 +1,515 @@
...
+```
+
+### Modules
 
-Prefer calling functions in the required scope vs. saving `this` in a `that` or `self` variable.
+Modules should be small, [composable](http://en.wikipedia.org/wiki/Composability), [stateless](http://en.wikipedia.org/wiki/State_(computer_science)), and self contained.
+
+#### Exporting
+
+Functions and primitive types shouldn't be default exports.
```
Comment (DevShep) [inline README.md]: Add an exception here: internal functions CAN be exported to a parent module iff (double 'f' intentional) they are entirely encapsulated by the parent.

```js
import a from 'func-a';
import b from 'func-b';

exports default {
  a: a,
  b: b
};
```

We care about the public API/interface of a module obeying this rule, not the internal implementation.

File: README.md [hunk 4]
```diff
@@ -1,222 +1,515 @@
...
+### Strict Mode
+
+* All modules should be have strict mode enabled.
+* ES6 modules have strict mode enabled by default.
+* Non-ES6 modules should have `'use strict';` placed at the top of the file.
+
+### Coerced Types and Equality
 
-Prefer `===` over `==`. This will be [enforced by JShint](http://www.jshint.com/docs/options/#eqeqeq).
+Strict comparison (`===`) should be used instead of abstract comparison (`==`).
```
Comment (DevShep) [inline README.md]: Exception for checking null? `foo != null` is equivalent to `foo !== null && foo !== undefined`. I think it's a good idiom to adopt and JSHINT has a flag for it.

File: README.md [hunk 5]
```diff
@@ -1,222 +1,515 @@
...
+Lines shouldn't have whitespace at the end.
 
-This is a simple way to catch mistyped variable names and copy/paste errors, but it also helps to prevent over-dependence on the global namespace, and forces modules to be more explicit about their dependencies.
+### 80 Column Limit
 
-One specific example of why this can be important is in Backbone views, which use a scoped `$` function to find elements inside the view. If you accidentally use the global `$('.selector')` instead of `this.$('.selector')`, you're searching the full DOM instead of just your view.
+Lines should not exceed 80 characters.
 
-## Using Variables Before Definition
+In specific cases, long lines can be ignored by the linter.
```
Comment (DevShep) [inline README.md]: This should either state to relax only the line-length rule or that the line that's ignored should only be a string. It's possible that ignoring an entire line means we miss valid linting concerns too.

**Comment by DevShep** [comment]
> Could we add (I guess only if everyone agrees on this) to enforce parens with ES6 arrow functions, even if there is only one parameter?

I don't think it's necessary. As I've been writing ES6, I've noticed that I don't normally add the extra params.
---

Can we add to prefer `'` to `"`? Both should be valid still, but I think it's custom that we prefer single to double quotes in new code.

File: README.md [hunk 6]
```diff
@@ -1,222 +1,515 @@
...
+### Functions
+
+Function expressions should be used instead of function declarations.
 
 **Don't do this:**
+
+```javascript
+function foo() {
+  /// [...]
+};
```
Comment (thealkimist) [inline README.md]: Even if we're not using them, function declarations shouldn't have a semi-colon.

File: README.md [hunk 7]
```diff
@@ -1,222 +1,515 @@
...
+  - [Strict Mode](#strict-mode)
+  - [Coerced Types and Equality](#coerced-types-and-equality)
+  - [Trailing Whitespace](#trailing-whitespace)
+  - [80 Column Limit](#80-column-limit)
+* [License](#license)
+
+## Introduction
+
+This guide outlines the style for Yola's JavaScript code.
+When editing existing code, and where possible, said code should be updated to conform to the style guide.
```
Comment (stefanor) [inline README.md]: That's a convoluted sentence. How about:

We do our best to update legacy code to conform to these standards, when editing it.

File: README.md [hunk 8]
```diff
@@ -1,222 +1,515 @@
...
+
+### Changes
+
+Proposed style changes should be discussed in a written format that can referenced.
+The style guide must be updated to reflect all changes.
```
Comment (stefanor) [inline README.md]: I think you're trying to say that any changes made to the linter config must be described in the style guide, too.

File: README.md [hunk 9]
```diff
@@ -1,222 +1,515 @@
...
+
+### Linting
+
+Much of this guide can be enforced by using a linter such as [JSHint](https://github.com/jshint/jshint).
+The provided [`.jshintrc`](https://github.com/yola/jsgreat/blob/master/.jshintrc) file should be used to help maintain conformity.
```
Comment (stefanor) [inline README.md]: Also rather an awkward sentence. Simplify!

File: README.md [hunk 10]
```diff
@@ -1,222 +1,515 @@
...
+let foo, bar, baz;
 ```
 
-## Function Declarations vs. Function Expressions
+#### Unused Variables
+
+Unused variables should be removed from the code when they are no longer needed.
 
-Prefer `var`ed function expressions over function declarations.
+Although discouraged, unused function parameters are allowed. This is due to some libraries expecting specific function signatures.
```
Comment (stefanor) [inline README.md]: Unused function parameters should be removed, unless necessary to maintain compatibility with other libraries. ?

File: README.md [hunk 11]
```diff
@@ -1,222 +1,515 @@
...
+let longTestPermalinkUrl;
+longTestPermalinkUrl = 'https://example.com/v1/en/2016/03/14/assets/images/1234.jpg'; // jshint ignore:line
+```
 
-## 80 Column Limit
+## License
 
-Although the style guide isn't fully explicit about it, lines should not exceed 80 characters.
+Copyright &copy; 2016 [Yola](http://yola.com).
+<br>Released under the [MIT License](https://github.com/yola/jsgreat/blob/master/LICENSE).
```
Comment (stefanor) [inline README.md]: It's better to be clear, and call it the MIT/Expat license. https://en.wikipedia.org/wiki/MIT_License#Various_versions

File: README.md [hunk 12]
```diff
@@ -1,222 +1,515 @@
...
-});
+if (foo) {
+  console.log('bar');
+}
 ```
 
-## Coerced Types and Equality
+### Strict Mode
+
+* All modules should be have strict mode enabled.
```
Comment (thealkimist) [inline README.md]: Feels like a redundancy.

File: README.md [hunk 13]
```diff
@@ -1,222 +1,515 @@
...
+Asynchronous operations should be written using promises.
+
+* Libraries reliant on promises should adhere to the [Promises/A+ standard]([Promises](http://promises-aplus.github.io/promises-spec/)).
+* [Bluebird](http://bluebirdjs.com) is Yola's promise implementation of choice.
+* ["You're missing the point of promises"](http://domenic.me/2012/10/14/youre-missing-the-point-of-promises/) by Domenic Denicola
+  - A valued article discussing the philosophy behind promises
+
+### JSDoc
+
+JSDoc shouldn't be used. Rules referring to JSDoc usage in the Google JavaScript Style Guide should be ignored.
```
Comment (thomasw) [inline README.md]: This isn't what I communicated here: https://github.com/yola/production/issues/3013#issuecomment-190867562

File: README.md (hunk 3 above)
Comment (DevShep) [inline README.md]: Because people don't know 'iff' is a thing.... https://en.wikipedia.org/wiki/If_and_only_if

Comment (thealkimist) [inline README.md]: And here I thought you were just implementing your own syntactic superlative for a level 2 'if'.

File: README.md (hunk 8 above)
Comment (morrisallison) [inline README.md]: That's not what I meant, but that's a good things to add.
I just wanted to add something to promote keeping the guide updated, but I should've just wrote that instead.

**Comment by morrisallison** [comment]
> Could we add (I guess only if everyone agrees on this) to enforce parens with ES6 arrow functions, even if there is only one parameter?

I'm of two minds about this, but generally, I agree.

I don't like these:

```javascript
arr.map((value) => foo + this.calulate(value));

fetchSomething()
  .then((res) => { return res.statusCode; });
```

I like these:

```javascript
arr.map((value) => {
  return foo + this.calulate(value);
});

fetchSomething()
  .then((res) => res.statusCode);
```

File: README.md (hunk 3 above)
Comment (beck) [inline README.md]: Unrelated to the point you're trying to make—shorthand notation?

```js
exports default { a, b };
```

File: README.md (hunk 2 above)
Comment (DevShep) [inline README.md]: > I do think that in promise chains, though, anonymous functions should not be allowed.

Fat arrow functions can be allowed, though.

```
promise
  .then(res => res.json())
  .then(somethingJSONy)
  .done
```

is fine imo

File: README.md [hunk 14]
```diff
@@ -1,222 +1,515 @@
...
+
+Extend the provided `.jshintrc` file:
+
+```json
+{
+  "extends": "./node_modules/jsgreat/.jshintrc"
+}
+```
+
+## Rules
```
Comment (beck) [inline README.md]: Suggest/preference: Guidelines

Rules should never be broken. Guidelines are allowed to be broken when circumstance demands.

File: README.md [hunk 15]
```diff
@@ -1,222 +1,515 @@
...
 
-Many of these options will be enforced by [this .jshintrc file](https://github.com/yola/wiki/blob/master/Files/.jshintrc). This file can be used by tools that support JSHint integration. For most projects, it is strongly recommended to use Grunt and the grunt-contrib-jshint plugin to run jshint.
+Serving as a base, the [Google JavaScript Style Guide](https://google.github.io/styleguide/javascriptguide.xml) should be referenced for topics not addressed below.
 
-The following additions and variations will be enforced:
+### Naming
 
-## Strings
-The guide [prefers `'` over `"`](http://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml?showone=Strings#Strings). We only require that quotes are used consistently within the file. This will be [enforced by jshint](http://www.jshint.com/docs/options/#quotmark).
+* Variables, constants, enums, methods, and properties should be written in `lowerCamelCase`.
```
Comment (morrisallison) [inline README.md]: @beck I'll update this to add that enum members should be uppercase.

**Comment by morrisallison** [comment]
Closing this. Iit was a mistake for me to lump all the changes together like this. I'll break these down into smaller chunks so we can discuss them separately.

**Pull request closed without merging**